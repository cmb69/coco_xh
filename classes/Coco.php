<?php

/**
 * Copyright 2012-2023 Christoph M. Becker
 *
 * This file is part of Coco_XH.
 *
 * Coco_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Coco_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Coco_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Coco;

use Coco\Infra\CsrfProtector;
use Coco\Infra\Repository;
use Coco\Infra\Request;
use Coco\Infra\XhStuff;
use Coco\Infra\View;
use Coco\Logic\Util;
use Coco\Value\Html;
use Coco\Value\Response;

class Coco
{
    /** @var Repository */
    private $repository;

    /** @var CsrfProtector */
    private $csrfProtector;

    /** @var XhStuff */
    private $xhStuff;

    /** @var View */
    private $view;

    public function __construct(
        Repository $repository,
        CsrfProtector $csrfProtector,
        XhStuff $xhStuff,
        View $view
    ) {
        $this->repository = $repository;
        $this->csrfProtector = $csrfProtector;
        $this->xhStuff = $xhStuff;
        $this->view = $view;
    }

    public function __invoke(Request $request, string $name, string $config, string $height): Response
    {
        if (!Util::isValidCocoName($name)) {
            return Response::create($this->view->message("fail", "error_invalid_name") . "\n");
        }
        if ($request->s() < 0) {
            return Response::create("");
        }
        switch ($request->cocoAction($name)) {
            default:
                return $this->show($request, $name);
            case "edit":
                return $this->edit($request, $name, $config, $height);
            case "do_edit":
                return $this->update($request, $name, $config, $height);
        }
    }

    private function show(Request $request, string $name): Response
    {
        $content = $this->repository->find($name, $request->s());
        $content = $this->xhStuff->evaluateScripting($content);
        $search = $request->search();
        if ($search !== "") {
            $words = Util::parseSearchTerm($search);
            $content = $this->xhStuff->highlightSearchWords($words, $content);
        }
        return Response::create($content);
    }

    private function edit(Request $request, string $name, string $config, string $height): Response
    {
        $content = $this->repository->find($name, $request->s());
        return Response::create($this->renderEditor($name, $config, $height, $content));
    }

    private function update(Request $request, string $name, string $config, string $height): Response
    {
        $this->csrfProtector->check();
        $text = $request->cocoText($name);
        if ($this->repository->save($name, $request->s(), $text)) {
            return Response::redirect($request->url()->absolute());
        }
        return Response::create(
            $this->view->message("fail", "error_save", $this->repository->filename($name))
            . $this->renderEditor($name, $config, $height, $text)
        );
    }

    private function renderEditor(string $name, string $config, string $height, string $content): string
    {
        $id = "coco_text_$name";
        $editor = $this->xhStuff->replaceEditor($id, $config);
        return $this->view->render("edit_form", [
            "id" => $id,
            "name" => $name,
            "style" => "width:100%; height:$height",
            "content" => $content,
            "editor" => $editor !== false ? Html::of($editor) : false,
            "csrf_token" => $this->csrfProtector->token(),
        ]);
    }
}
