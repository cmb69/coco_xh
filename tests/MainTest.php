<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Coco_XH.
 *
 * Coco_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Coco_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Coco_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Coco;

use ApprovalTests\Approvals;
use Coco\Infra\Repository;
use Coco\Infra\RequestStub;
use Coco\Infra\View;
use PHPUnit\Framework\TestCase;

class MainTest extends TestCase
{
    public function testDoesNothingWhenNotLoggingOut(): void
    {
        $sut = new Main($this->conf(), $this->repository(), $this->view());
        $response = $sut(new RequestStub());
        $this->assertEquals("", $response->output());
    }

    public function testReportsBackupSuccess(): void
    {
        $sut = new Main($this->conf(), $this->repository(), $this->view());
        $request = new RequestStub(["time" => strtotime("2023-03-06T12:00:00"), "logOut" => true]);
        $response = $sut($request);
        Approvals::verifyHtml($response->output());
    }

    public function testReportsFailureToCreateBackup(): void
    {
        $sut = new Main($this->conf(), $this->repository(false), $this->view());
        $request = new RequestStub(["time" => strtotime("2023-03-06T12:00:00"), "logOut" => true]);
        $response = $sut($request);
        Approvals::verifyHtml($response->output());
    }

    public function testReportsFailureToDeleteSurplusBackups(): void
    {
        $sut = new Main($this->conf(), $this->repository(true, false), $this->view());
        $request = new RequestStub(["time" => strtotime("2023-03-06T12:00:00"), "logOut" => true]);
        $response = $sut($request);
        Approvals::verifyHtml($response->output());
    }

    private function repository(bool $create = true, bool $delete = true): Repository
    {
        $repository = $this->createStub(Repository::class);
        $repository->method("dataFolder")->willReturn("./content/coco/");
        $repository->method("findAllNames")->willReturn(["foo", "bar"]);
        $repository->method("filename")->willReturnCallback(function ($coconame, $date = null) {
            return "./content/coco/" . ($date !== null ? "{$date}_" : "") . "$coconame.htm";
        });
        $repository->method("backup")->willReturn($create);
        $repository->method("delete")->willReturn($delete);
        $repository->method("findAllBackups")->willReturnOnConsecutiveCalls([
            ["foo", "20230304_120000"],
            ["foo", "20230305_120000"],
            ["foo", "20230306_120000"],
        ], [
            ["bar", "20230304_120000"],
            ["bar", "20230305_120000"],
            ["bar", "20230306_120000"],
        ]);
        return $repository;
    }

    private function view(): View
    {
        return new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["coco"]);
    }

    private function conf(): array
    {
        return ["backup_numberoffiles" => 2];
    }
}
