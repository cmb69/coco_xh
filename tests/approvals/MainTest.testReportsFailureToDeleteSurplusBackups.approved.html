<p class="xh_info">./content/coco/20230306_120000_foo.htm created</p>
<p class="xh_fail">./content/coco/20230304_120000_foo.htm could not be deleted!</p>
<p class="xh_info">./content/coco/20230306_120000_bar.htm created</p>
<p class="xh_fail">./content/coco/20230304_120000_bar.htm could not be deleted!</p>
